src/filelock/_api.py

```diff
@@ -52,29 +89,30 @@ def __init__(
         :param timeout: default timeout when acquiring the lock, in seconds. It will be used as fallback value in
         the acquire method, if no timeout value (``None``) is given. If you want to disable the timeout, set it
         to a negative value. A timeout of 0 means, that there is exactly one attempt to acquire the file lock.
-        : param mode: file permissions for the lockfile.
+        :param mode: file permissions for the lockfile.
+        :param thread_local: Whether this object's internal context should be thread local or not.
+        If this is set to ``False`` then the lock will be rentrant across threads.
         """
-        # The path to the lock file.
-        self._lock_file: str = os.fspath(lock_file)
-
-        # The file descriptor for the *_lock_file* as it is returned by the os.open() function.
-        # This file lock is only NOT None, if the object currently holds the lock.
-        self._lock_file_fd: int | None = None
-
-        # The default timeout value.
-        self._timeout: float = timeout
-
-        # The mode for the lock files
-        self._mode: int = mode
-
-        # The lock counter is used for implementing the nested locking mechanism. Whenever the lock is acquired, the
-        # counter is increased and the lock is only released, when this value is 0 again.
-        self._lock_counter: int = 0
+        # Whether or not this object is thread local or not
+        self.thread_local = thread_local
+
+        # Create the context. Note that external code should not work with the context directly
+        # and should instead use properties of this class.
+        context_kwargs: dict[str, Any] = {
+            "lock_file": os.fspath(lock_file),
+            "timeout": timeout,
+            "mode": mode,
+        }
+        self._context: FileLockContext | ThreadLocalFileContext
```

Comment:
```suggestion
        self._context: FileLockContext
```
should be enough, because `ThreadLocalFileContext` is also a `FileLockContext`.

I think `mypy` complained because the "first" branch assigned an instance of the subclass, and may not have complained if the `if ... else` branches would have been swapped (not suggesting that, though).